File: frontend/scripts/generate-tokens-css.py
#!/usr/bin/env python3
"""Генерирует frontend/css/tokens.css из design/tokens.json.

Источник истины — design/tokens.json (общий контракт web/Android). Запускать заново
после любого изменения tokens.json:

    python3 frontend/scripts/generate-tokens-css.py

Правила именования CSS custom properties:
  color.light.background.canvas   -> --color-background-canvas          (на :root, светлая тема)
  color.dark.background.canvas    -> --color-background-canvas          (на :root[data-theme="dark"])
  spacing.4                       -> --spacing-4
  radius.sm                       -> --radius-sm
  elevation.1                     -> --shadow-1
  motion.durationFast             -> --motion-duration-fast
  typography.scale.h1.size        -> --font-size-h1 (rem), --line-height-h1, --font-weight-h1
"""
import json
import pathlib
import re

ROOT = pathlib.Path(__file__).resolve().parents[2]
TOKENS_PATH = ROOT / "design" / "tokens.json"
OUT_PATH = ROOT / "frontend" / "css" / "tokens.css"


def kebab(name):
    s = re.sub(r"([a-z0-9])([A-Z])", r"\1-\2", name)
    return s.lower()


def color_props(theme_dict):
    lines = []
    for group, values in theme_dict.items():
        if group.startswith("_"):
            continue
        for key, val in values.items():
            if key.startswith("_"):
                continue
            lines.append(f"  --color-{kebab(group)}-{kebab(key)}: {val};")
    return lines


def main():
    tokens = json.loads(TOKENS_PATH.read_text(encoding="utf-8"))

    light_lines = color_props(tokens["color"]["light"])
    dark_lines = color_props(tokens["color"]["dark"]) if "dark" in tokens["color"] else []

    spacing_lines = [
        f"  --spacing-{k}: {v}px;" for k, v in tokens["spacing"].items() if not k.startswith("_")
    ]
    radius_lines = [
        f"  --radius-{kebab(k)}: {v if isinstance(v, str) else str(v) + 'px'};"
        for k, v in tokens["radius"].items()
        if not k.startswith("_")
    ]
    shadow_lines = [
        f"  --shadow-{k}: {v};" for k, v in tokens["elevation"].items() if not k.startswith("_")
    ]
    motion = tokens["motion"]
    motion_lines = [
        f"  --motion-duration-fast: {motion['durationFast']}ms;",
        f"  --motion-duration-base: {motion['durationBase']}ms;",
        f"  --motion-duration-slow: {motion['durationSlow']}ms;",
        f"  --motion-easing-standard: {motion['easingStandard']};",
        f"  --motion-easing-emphasized: {motion['easingEmphasized']};",
    ]

    typo = tokens["typography"]
    font_lines = [
        f"  --font-family-base: {typo['fontFamily']['base']};",
        f"  --font-family-mono: {typo['fontFamily']['mono']};",
    ]
    for name, scale in typo["scale"].items():
        key = kebab(name)
        font_lines.append(f"  --font-size-{key}: {scale['size']}rem;")
        font_lines.append(f"  --line-height-{key}: {scale['lineHeight']};")
        font_lines.append(f"  --font-weight-{key}: {scale['weight']};")
        if "fontStyle" in scale:
            font_lines.append(f"  --font-style-{key}: {scale['fontStyle']};")

    out = []
    out.append("/* СГЕНЕРИРОВАНО из design/tokens.json скриптом")
    out.append("   frontend/scripts/generate-tokens-css.py — не редактировать вручную,")
    out.append("   при изменении tokens.json перезапустить скрипт. */")
    out.append("")
    out.append(":root {")
    out.extend(light_lines)
    out.append("")
    out.extend(spacing_lines)
    out.extend(radius_lines)
    out.extend(shadow_lines)
    out.extend(motion_lines)
    out.extend(font_lines)
    out.append("}")

    if dark_lines:
        out.append("")
        out.append('/* Тёмная тема — активируется через :root[data-theme="dark"] (переключатель в Профиле). */')
        out.append('@media (prefers-color-scheme: dark) {')
        out.append('  :root:not([data-theme="light"]) {')
        out.extend("  " + l for l in dark_lines)
        out.append("  }")
        out.append("}")
        out.append(':root[data-theme="dark"] {')
        out.extend(dark_lines)
        out.append("}")

    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    OUT_PATH.write_text("\n".join(out) + "\n", encoding="utf-8")
    print(f"written {OUT_PATH} ({len(out)} lines)")


if __name__ == "__main__":
    main()

File: frontend/scripts/generate-tokens-css.mjs
#!/usr/bin/env node
/**
 * Генерирует frontend/src/styles/tokens.css из design/tokens.json.
 *
 * Источник истины — design/tokens.json (общий контракт web/Android), который сам не входит
 * в репозиторий фронтенда. Скрипт запускается автоматически перед dev/build (npm pre-хуки),
 * должен работать на голом Node без зависимостей. Если design/tokens.json недоступен (например,
 * сборка выполняется без соседнего репозитория дизайна) — скрипт молча ничего не делает и
 * оставляет уже закоммиченный frontend/src/styles/tokens.css как есть.
 *
 * Правила именования CSS custom properties — см. исходную версию на Python
 * (frontend/scripts/generate-tokens-css.py в истории), сохранены без изменений:
 *   color.light.background.canvas -> --color-background-canvas          (:root, светлая тема)
 *   color.dark.background.canvas  -> --color-background-canvas          (:root[data-theme="dark"])
 *   spacing.4                     -> --spacing-4
 *   radius.sm                     -> --radius-sm
 *   elevation.1                   -> --shadow-1
 *   motion.durationFast           -> --motion-duration-fast
 *   typography.scale.h1.size      -> --font-size-h1 (rem), --line-height-h1, --font-weight-h1
 */
import { readFileSync, writeFileSync, mkdirSync, existsSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..", "..");
const TOKENS_PATH = path.join(ROOT, "design", "tokens.json");
const OUT_PATH = path.join(__dirname, "..", "src", "styles", "tokens.css");

function kebab(name) {
  return name.replace(/([a-z0-9])([A-Z])/g, "$1-$2").toLowerCase();
}

function colorProps(themeDict) {
  const lines = [];
  for (const [group, values] of Object.entries(themeDict)) {
    if (group.startsWith("_")) continue;
    for (const [key, val] of Object.entries(values)) {
      if (key.startsWith("_")) continue;
      lines.push(`  --color-${kebab(group)}-${kebab(key)}: ${val};`);
    }
  }
  return lines;
}

function main() {
  if (!existsSync(TOKENS_PATH)) {
    console.log(`design/tokens.json not found (${TOKENS_PATH}) — keeping existing tokens.css as is`);
    return;
  }

  let tokens;
  try {
    tokens = JSON.parse(readFileSync(TOKENS_PATH, "utf-8"));
  } catch (err) {
    console.warn(`could not read/parse design/tokens.json (${err.message}) — keeping existing tokens.css as is`);
    return;
  }

  const lightLines = colorProps(tokens.color.light);
  const darkLines = tokens.color.dark ? colorProps(tokens.color.dark) : [];

  const spacingLines = Object.entries(tokens.spacing)
    .filter(([k]) => !k.startsWith("_"))
    .map(([k, v]) => `  --spacing-${k}: ${v}px;`);

  const radiusLines = Object.entries(tokens.radius)
    .filter(([k]) => !k.startsWith("_"))
    .map(([k, v]) => `  --radius-${kebab(k)}: ${typeof v === "string" ? v : `${v}px`};`);

  const shadowLines = Object.entries(tokens.elevation)
    .filter(([k]) => !k.startsWith("_"))
    .map(([k, v]) => `  --shadow-${k}: ${v};`);

  const motion = tokens.motion;
  const motionLines = [
    `  --motion-duration-fast: ${motion.durationFast}ms;`,
    `  --motion-duration-base: ${motion.durationBase}ms;`,
    `  --motion-duration-slow: ${motion.durationSlow}ms;`,
    `  --motion-easing-standard: ${motion.easingStandard};`,
    `  --motion-easing-emphasized: ${motion.easingEmphasized};`
  ];

  const typo = tokens.typography;
  const fontLines = [
    `  --font-family-base: ${typo.fontFamily.base};`,
    `  --font-family-mono: ${typo.fontFamily.mono};`
  ];
  for (const [name, scale] of Object.entries(typo.scale)) {
    const key = kebab(name);
    fontLines.push(`  --font-size-${key}: ${scale.size}rem;`);
    fontLines.push(`  --line-height-${key}: ${scale.lineHeight};`);
    fontLines.push(`  --font-weight-${key}: ${scale.weight};`);
    if (scale.fontStyle) fontLines.push(`  --font-style-${key}: ${scale.fontStyle};`);
  }

  const out = [];
  out.push("/* СГЕНЕРИРОВАНО из design/tokens.json скриптом");
  out.push("   frontend/scripts/generate-tokens-css.mjs — не редактировать вручную,");
  out.push("   при изменении tokens.json перезапустить скрипт (npm run dev / npm run build делают это сами). */");
  out.push("");
  out.push(":root {");
  out.push(...lightLines);
  out.push("");
  out.push(...spacingLines);
  out.push(...radiusLines);
  out.push(...shadowLines);
  out.push(...motionLines);
  out.push(...fontLines);
  out.push("}");

  if (darkLines.length > 0) {
    out.push("");
    out.push('/* Тёмная тема — активируется через :root[data-theme="dark"] (переключатель в Профиле). */');
    out.push("@media (prefers-color-scheme: dark) {");
    out.push('  :root:not([data-theme="light"]) {');
    out.push(...darkLines.map((l) => "  " + l));
    out.push("  }");
    out.push("}");
    out.push(':root[data-theme="dark"] {');
    out.push(...darkLines);
    out.push("}");
  }

  mkdirSync(path.dirname(OUT_PATH), { recursive: true });
  writeFileSync(OUT_PATH, out.join("\n") + "\n", "utf-8");
  console.log(`written ${OUT_PATH} (${out.length} lines)`);
}

main();
